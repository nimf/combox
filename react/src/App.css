.App {
  max-width: 600px;
  margin: 0 auto;
}

.author .ui.mini.label {
  top: -0.2em;
  position: relative;
  padding: .6em .833em .4em;
  margin-left: .8em;
}
