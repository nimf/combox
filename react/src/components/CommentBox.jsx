import React from 'react';
import PropTypes from 'prop-types';
import { Form, Button } from 'semantic-ui-react';

const propTypes = {
  text: PropTypes.string,
  onChange: PropTypes.func.isRequired,
  onSubmit: PropTypes.func.isRequired,
  enabled: PropTypes.bool.isRequired,
};
const defaultProps = {
  text: '',
};

function CommentBox({
  text, onChange, onSubmit, enabled,
}) {
  return (
    <Form onSubmit={onSubmit}>
      <Form.TextArea
        autoHeight
        rows={2}
        value={text}
        onChange={onChange}
        placeholder="Leave a comment..."
      />
      <Button
        disabled={!enabled || text.trim().length === 0}
        content="Add Comment"
        labelPosition="left"
        icon="edit"
        primary
      />
    </Form>
  );
}

CommentBox.propTypes = propTypes;
CommentBox.defaultProps = defaultProps;

export default CommentBox;
