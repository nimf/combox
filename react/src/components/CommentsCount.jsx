import React from 'react';
import PropTypes from 'prop-types';
import { Header } from 'semantic-ui-react';

const propTypes = {
  count: PropTypes.number,
};

const defaultProps = {
  count: undefined,
};

function CommentsCount({ count }) {
  let displayCount = 'Loading comments...';
  if (count !== undefined) {
    displayCount = count > 0
      ? `${count} comment${count > 1 ? 's' : ''}`
      : 'No comments yet';
  }
  return <Header as="h4">{displayCount}</Header>;
}

CommentsCount.propTypes = propTypes;
CommentsCount.defaultProps = defaultProps;

export default CommentsCount;
