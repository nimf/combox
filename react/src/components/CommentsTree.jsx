import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Comment } from 'semantic-ui-react';
import CommentNode from './CommentNode';

const propTypes = {
  comments: PropTypes.object.isRequired,
  currentCommentId: PropTypes.number,
  onFocusedRendered: PropTypes.func.isRequired,
  onOpenNewComments: PropTypes.func.isRequired,
  onToggleReply: PropTypes.func.isRequired,
  onReplyChange: PropTypes.func.isRequired,
  onReply: PropTypes.func.isRequired,
};
const defaultProps = {
  currentCommentId: null,
};

class CommentsTree extends Component {
  componentDidUpdate() {
    if (this.props.comments.focusedId !== null &&
        this.props.comments.byId[this.props.comments.focusedId].parentId ===
        this.props.currentCommentId) {
      // If we rendered a comment that must receive focus, report it
      this.props.onFocusedRendered();
    }
  }

  onOpenNewComments = () =>
    this.props.onOpenNewComments(this.props.currentCommentId);

  onToggleReply = commentId =>
    () => this.props.onToggleReply(commentId);

  onReplyChange = commentId =>
    event => this.props.onReplyChange(commentId, event);

  onReply = commentId =>
    () => this.props.onReply(commentId);

  render() {
    const currentComments = this.props.comments.allIds.map(id => this.props.comments.byId[id])
      .filter(c => c.parentId === this.props.currentCommentId);
    const hiddenCount = currentComments.filter(c => c.hidden).length;
    if (currentComments.length === 0) return null;
    const listOfComments = currentComments.filter(c => !c.hidden).map(comment =>
      (
        <CommentNode
          key={comment.id}
          comment={comment}
          fresh={this.props.comments.focusedId === comment.id}
          onToggleReply={this.onToggleReply(comment.id)}
          onReplyChange={this.onReplyChange(comment.id)}
          onReply={this.onReply(comment.id)}
        >
          <CommentsTree
            comments={this.props.comments}
            currentCommentId={comment.id}
            onFocusedRendered={this.props.onFocusedRendered}
            onOpenNewComments={this.props.onOpenNewComments}
            onToggleReply={this.props.onToggleReply}
            onReplyChange={this.props.onReplyChange}
            onReply={this.props.onReply}
          />
        </CommentNode>
      ));
    return (
      <Comment.Group threaded={this.props.currentCommentId === null}>
        {listOfComments}
        {hiddenCount > 0 &&
          <Comment>
            <Comment.Content>
              <Comment.Author
                as="a"
                onClick={
                  this.onOpenNewComments
                }
              >
                {hiddenCount > 1
                  ? `${hiddenCount} new comments`
                  : '1 new comment'
                }
              </Comment.Author>
            </Comment.Content>
          </Comment>
        }
      </Comment.Group>
    );
  }
}

CommentsTree.propTypes = propTypes;
CommentsTree.defaultProps = defaultProps;

export default CommentsTree;
