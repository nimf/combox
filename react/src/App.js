import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Segment } from 'semantic-ui-react';
import 'scroll-behaviour/polyfill';
import { connectToChannel } from './actions/channel';
import {
  saveDraft, postComment, focusComment, openNewComments, toggleReply,
  replyChanged, reply,
} from './actions/subject';
import CommentBox from './components/CommentBox';
import CommentsCount from './components/CommentsCount';
import CommentsTree from './components/CommentsTree';
import './App.css';

const propTypes = {
  connectToChannel: PropTypes.func.isRequired,
  saveDraft: PropTypes.func.isRequired,
  postComment: PropTypes.func.isRequired,
  focusComment: PropTypes.func.isRequired,
  openNewComments: PropTypes.func.isRequired,
  channelConnected: PropTypes.bool.isRequired,
  toggleReply: PropTypes.func.isRequired,
  replyChanged: PropTypes.func.isRequired,
  reply: PropTypes.func.isRequired,
  displayName: PropTypes.string,
  channel: PropTypes.object,
  commentsCount: PropTypes.number,
  draftComment: PropTypes.string,
  comments: PropTypes.object.isRequired,
};
const defaultProps = {
  channel: undefined,
  commentsCount: undefined,
  draftComment: '',
  displayName: 'Anonymous',
};

export class App extends Component {
  componentDidMount() {
    this.props.connectToChannel();
  }

  onPostComment = () => {
    this.props.postComment(
      this.props.channel,
      this.props.displayName,
      this.props.draftComment,
    );
  }

  onSaveDraft = (event) => {
    this.props.saveDraft(event.target.value);
  }

  onFocusedRendered = () => {
    this.props.focusComment(this.props.comments.focusedId);
  }

  onOpenNewComments = (commentId) => {
    this.props.openNewComments(commentId);
  }

  onToggleReply = (commentId) => {
    this.props.toggleReply(commentId);
  }

  onReplyChange = (commentId, event) => {
    this.props.replyChanged(commentId, event.target.value);
  }

  onReply = (commentId) => {
    this.props.reply(commentId);
  }

  render() {
    return (
      <div className="App">
        <Segment basic>
          <CommentsCount count={this.props.commentsCount} />
          <CommentBox
            text={this.props.draftComment}
            enabled={this.props.channelConnected}
            onChange={this.onSaveDraft}
            onSubmit={this.onPostComment}
          />
          <CommentsTree
            comments={this.props.comments}
            onFocusedRendered={this.onFocusedRendered}
            onOpenNewComments={this.onOpenNewComments}
            onToggleReply={this.onToggleReply}
            onReplyChange={this.onReplyChange}
            onReply={this.onReply}
          />
        </Segment>
      </div>
    );
  }
}
App.propTypes = propTypes;
App.defaultProps = defaultProps;

export default connect(
  state => ({
    channelConnected: state.channel.channel !== undefined,
    channel: state.channel.channel,
    commentsCount: state.subject.commentsCount,
    draftComment: state.subject.draftComment,
    displayName: state.user.displayName,
    comments: state.subject.comments,
  }),
  {
    connectToChannel,
    saveDraft,
    postComment,
    focusComment,
    openNewComments,
    toggleReply,
    replyChanged,
    reply,
  },
)(App);
