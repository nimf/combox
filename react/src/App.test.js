import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import { App } from './App';
import CommentBox from './components/CommentBox';
import CommentsCount from './components/CommentsCount';
import CommentsTree from './components/CommentsTree';
import commentsFixture from './fixtures/comments';

test('App renders correctly', () => {
  const tree = renderer.create(<App
    connectToChannel={() => {}}
    saveDraft={() => {}}
    postComment={() => {}}
    openNewComments={() => {}}
    focusComment={() => {}}
    channelConnected={false}
    comments={commentsFixture}
  />).toJSON();
  expect(tree).toMatchSnapshot();
});

describe('App Shallow Render', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<App
      connectToChannel={() => {}}
      saveDraft={() => {}}
      postComment={() => {}}
      openNewComments={() => {}}
      focusComment={() => {}}
      channelConnected={false}
      comments={{ byId: {}, allIds: [] }}
    />);
  });

  it('renders the component', () => {
    expect(wrapper.length).toEqual(1);
  });

  it('renders CommentsCount, CommentBox and CommentsTree', () => {
    expect(wrapper.find(CommentsCount).length).toEqual(1);
    expect(wrapper.find(CommentBox).length).toEqual(1);
    expect(wrapper.find(CommentsTree).length).toEqual(1);
  });
});
