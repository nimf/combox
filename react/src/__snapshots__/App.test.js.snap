// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
<div
  className="App"
>
  <div
    className="ui basic segment"
  >
    <h4
      className="ui header"
    >
      Loading comments...
    </h4>
    <form
      action={undefined}
      className="ui form"
      onSubmit={[Function]}
    >
      <div
        className="field"
      >
        <textarea
          content={undefined}
          disabled={undefined}
          onChange={[Function]}
          onInput={[Function]}
          placeholder="Leave a comment..."
          required={undefined}
          rows={2}
          style={
            Object {
              "resize": "none",
            }
          }
          type={undefined}
          value=""
        />
      </div>
      <button
        className="ui icon primary disabled left labeled button"
        disabled={true}
        onClick={[Function]}
        role="button"
        tabIndex={-1}
      >
        <i
          aria-hidden="true"
          className="edit icon"
        />
        Add Comment
      </button>
    </form>
    <div
      className="ui threaded comments"
    >
      <div
        className="comment"
        id="comment-1"
      >
        <a
          className="avatar"
        >
          <img
            src="https://www.gravatar.com/avatar/00000000000000000000000000000000?s=100"
          />
        </a>
        <div
          className="content"
        >
          <span
            className="author"
          >
            <span>
              Anonymous
            </span>
            <div
              className="ui mini label"
              onClick={[Function]}
            >
              Guest
            </div>
          </span>
          <div
            className="metadata"
          >
            <div
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
            >
              3 days ago
            </div>
          </div>
          <div
            className="text"
          >
            <p>
              Test comment 1
            </p>
          </div>
          <div
            className="actions"
          >
            <a
              className=""
              onClick={[Function]}
            >
              Reply
            </a>
          </div>
        </div>
        <div
          className="ui comments"
        >
          <div
            className="comment"
            id="comment-2"
          >
            <a
              className="avatar"
            >
              <img
                src="https://www.gravatar.com/avatar/00000000000000000000000000000000?s=100"
              />
            </a>
            <div
              className="content"
            >
              <span
                className="author"
              >
                <span>
                  Anonymous
                </span>
                <div
                  className="ui mini label"
                  onClick={[Function]}
                >
                  Guest
                </div>
              </span>
              <div
                className="metadata"
              >
                <div
                  onBlur={[Function]}
                  onClick={[Function]}
                  onFocus={[Function]}
                  onMouseEnter={[Function]}
                  onMouseLeave={[Function]}
                >
                  2 days ago
                </div>
              </div>
              <div
                className="text"
              >
                <p>
                  Test comment 2
                </p>
              </div>
              <div
                className="actions"
              >
                <a
                  className=""
                  onClick={[Function]}
                >
                  Reply
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="comment"
        id="comment-3"
      >
        <a
          className="avatar"
        >
          <img
            src="https://www.gravatar.com/avatar/00000000000000000000000000000000?s=100"
          />
        </a>
        <div
          className="content"
        >
          <span
            className="author"
          >
            <span>
              Anonymous
            </span>
            <div
              className="ui mini label"
              onClick={[Function]}
            >
              Guest
            </div>
          </span>
          <div
            className="metadata"
          >
            <div
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
            >
              a day ago
            </div>
          </div>
          <div
            className="text"
          >
            <p>
              Test comment 3
            </p>
          </div>
          <div
            className="actions"
          >
            <a
              className=""
              onClick={[Function]}
            >
              Reply
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
